### PR TITLE
Add Azure Storage Queues Transport target

### PR DESCRIPTION
**Pull request by WilliamBZA**

### CODE REVIEW

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+            // Assert
+            Assert.IsFalse(ableToMigrate.CanMigrate);
+            Assert.AreEqual("Unable to connect to the storage instance with connection string 'DefaultEndpointsProtocol=https;AccountName=fakename;AccountKey=g94OvNO9o3sVan5eipemQEHmU8zD2M9iq98E8nKSdR2bTuQB1hi07Yd1/8dDw6+1jGI2klWjpvoDahHPhR/3og=='. Exception message 'No such host is known.'", ableToMigrate.Problems[0]);
+        }
+
+        [Test]
+        public async Task AbleToMigrate_passes_if_Timeout_table_exists()
+        {
+            // Arrange
+            await CreateTimeoutTable($"T{endpointName.Replace("-", string.Empty)}").ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:65]: Why do you need to replace here again. Doesn't the endpoint name already have the `-` replaced?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+        }
+
+        [Test]
+        public async Task StageBatch_inserts_timeouts_into_the_staging_queue()
+        {
+            // Arrange
+            var nameProvider = new DelayedDeliveryTableNameProvider();
+
+            var timeoutTarget = new ASQTarget(new TestLoggingAdapter(), connectionString, nameProvider);
+            var migrator = await timeoutTarget.PrepareTargetEndpointBatchMigrator(endpointName).ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:97]: await using?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        [Test]
+        public async Task CompleteBatch_Moves_All_Entries_From_Staging_To_DelayedMessageTable()
+        {
+            // Arrange
+            var nameProvider = new DelayedDeliveryTableNameProvider();
+
+            await CreateTimeoutTable(nameProvider.GetDelayedDeliveryTableName(endpointName)).ConfigureAwait(true);
+
+            var timeoutTarget = new ASQTarget(new TestLoggingAdapter(), connectionString, nameProvider);
+            var migrator = await timeoutTarget.PrepareTargetEndpointBatchMigrator(endpointName).ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:142]: await using?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+        [Test]
+        public async Task CompleteBatch_Removes_All_Entries_From_Staging()
+        {
+            // Arrange
+            var nameProvider = new DelayedDeliveryTableNameProvider();
+
+            await CreateTimeoutTable(nameProvider.GetDelayedDeliveryTableName(endpointName)).ConfigureAwait(true);
+
+            var timeoutTarget = new ASQTarget(new TestLoggingAdapter(), connectionString, nameProvider);
+            var migrator = await timeoutTarget.PrepareTargetEndpointBatchMigrator(endpointName).ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:189]: await using?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+        [Test]
+        public async Task Abort_Removes_Staging_Table()
+        {
+            // Arrange
+            var nameProvider = new DelayedDeliveryTableNameProvider();
+
+            await CreateTimeoutTable(nameProvider.GetDelayedDeliveryTableName(endpointName)).ConfigureAwait(true);
+
+            var timeoutTarget = new ASQTarget(new TestLoggingAdapter(), connectionString, nameProvider);
+            await timeoutTarget.PrepareTargetEndpointBatchMigrator(endpointName).ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:235]: await using?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+        [Test]
+        public async Task Complete_Removes_Staging_Queue_If_Empty()
+        {
+            // Arrange
+            var nameProvider = new DelayedDeliveryTableNameProvider();
+
+            await CreateTimeoutTable(nameProvider.GetDelayedDeliveryTableName(endpointName)).ConfigureAwait(true);
+
+            var timeoutTarget = new ASQTarget(new TestLoggingAdapter(), connectionString, nameProvider);
+            var migrator = await timeoutTarget.PrepareTargetEndpointBatchMigrator(endpointName).ConfigureAwait(true);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:300]: await using?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+
+            // Assert
+            var stagingTableExists = await DoesTableExist(nameProvider.GetStagingTableName(endpointName)).ConfigureAwait(true);
+
+            Assert.IsFalse(stagingTableExists);
+        }
+
+        async Task DeleteTable(string tableName)
+        {
+            var cloudStorageAccount = CloudStorageAccount.Parse(connectionString);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:341]: It doesn't really matter that much but since we have a dedicated setup and teardown phase would it make sense to create the client once, assign it to a field and then just reuse that?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/DelayedDeliveryTableNameProviderTests.cs [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+        public void When_Migrating_Endpoint_With_DelayedDeliveryTableOverride_DelayedDeliveryTableName_Is_NotComputed()
+        {
+            // Arrange
+            var delayedDeliveryTableNameGenerator = new DelayedDeliveryTableNameProvider("overriden delayed delivery table name");
+
+            // Act
+            var delayedDeliveryTableName = delayedDeliveryTableNameGenerator.GetDelayedDeliveryTableName("EndpointName");
+
+            // Assert
+            Assert.AreEqual("overriden delayed delivery table name", delayedDeliveryTableName);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/DelayedDeliveryTableNameProviderTests.cs:36]: That doesn't seem to be a valid azure table name, is that OK?

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/TimeoutMigrationTool.ASQ.IntegrationTests.csproj [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+    <IsPackable>false</IsPackable>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="NUnit" Version="3.13.0" />
+    <PackageReference Include="NUnit3TestAdapter" Version="3.17.0" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.8.3" />
+    <PackageReference Include="Particular.CodeRules" Version="0.8.0" PrivateAssets="All" />
+    <PackageReference Include="Particular.Approvals" Version="0.2.0" />
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.ASQ.IntegrationTests/TimeoutMigrationTool.ASQ.IntegrationTests.csproj:14]: Isn't the guidance to explicitly reference the third-party SDKs in the tests instead of using transitive dependencies? 

https://github.com/Particular/PlatformArchitecture/blob/6bf2a2359a4c6588daf6ef04cc1a8cb2dce1f9a5/repo-standards/README.md#test-projects-1

You could argue it doesn't matter that much since we are not using ranges in the tool

File: src/TimeoutMigrationTool.NHibernate.AcceptanceTests/TimeoutMigrationTool.NHibernate.AcceptanceTests.csproj [hunk 10]
```diff
@@ -8,8 +8,10 @@
 
   <ItemGroup>
     <PackageReference Include="NServiceBus.AcceptanceTesting" Version="7.4.4" />
-    <PackageReference Include="NServiceBus.RabbitMQ" Version="6.0.0" />
+    <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.2.0" />
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.NHibernate.AcceptanceTests/TimeoutMigrationTool.NHibernate.AcceptanceTests.csproj:11]: Same comment as above regarding the package references

File: src/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests.csproj [hunk 11]
```diff
@@ -10,12 +10,15 @@
...
     <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\EndpointCustomizationConfigurationExtensions.cs" Link="EndpointCustomizationConfigurationExtensions.cs" />
     <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\NHibernateTimeoutsSourceTests.cs" Link="NHibernateTimeoutsSourceTests.cs" />
+    <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\NHibernateToASQEndToEnd.cs" Link="NHibernateToASQEndToEnd.cs" />
     <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\NHibernateToRabbitMqEndToEnd.cs" Link="NHibernateToRabbitMqEndToEnd.cs" />
     <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\DefaultServer.cs" Link="DefaultServer.cs" />
     <Compile Include="..\TimeoutMigrationTool.NHibernate.AcceptanceTests\TestLoggingAdapter.cs" Link="TestLoggingAdapter.cs" />
   </ItemGroup>
 
     <ItemGroup>
+      <PackageReference Include="NServiceBus.Newtonsoft.Json" Version="2.2.0" />
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests.csproj:20]: Here as well

File: src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+
+                        await WaitUntilTheTimeoutIsSavedInRaven(ravenAdapter, sourceEndpoint);
+
+                        c.TimeoutSet = true;
+                    }))
+                .Done(c => c.TimeoutSet)
+                .Run(TimeSpan.FromSeconds(15));
+
+            var context = await Scenario.Define<TargetTestContext>()
+                .WithEndpoint<LegacyRavenDBEndpoint>(b => b.CustomConfig(ec =>
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs:56]: Why is it necessary to bootstrap this one as well?

File: src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPAcceptanceTest.cs [hunk 13]
```diff
@@ -30,7 +30,7 @@ public async Task SetUp()
 
                 testName = testName.Replace("_", "");
 
-                return testName + "_" + endpointBuilder;
+                return testName + "-" + endpointBuilder;
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPAcceptanceTest.cs:33]: Ok so this is that we get valid endpoint name prefixes for ASQ?

File: src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+        {
+            public bool TimeoutSet { get; set; }
+        }
+
+        public class TargetTestContext : ScenarioContext
+        {
+            public bool GotTheDelayedMessage { get; set; }
+        }
+
+        public class SalesSqlPEndpoint : EndpointConfigurationBuilder
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs:132]: Why are we suddenly using fancy names here?

File: src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+
+                    await migrationRunner.Run(DateTime.Now.AddDays(-10), EndpointFilter.SpecificEndpoint(salesEndpoint), new Dictionary<string, string>());
+                }))
+                .Done(c => c.GotTheDelayedMessage)
+                .Run(TimeSpan.FromSeconds(30));
+
+            Assert.True(context.GotTheDelayedMessage);
+        }
+
+        async Task WaitUntilTheTimeoutIsSavedInSql(string endpoint)
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs:109]: Should this be in the base class?

File: src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs (hunk 12 above)
Comment (danielmarbach) [reply src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs:56]: Is it because of https://github.com/Particular/TimeoutMigrationTool/pull/171/files#diff-e74c55f9dbb0e389c88bfe007b80d5d37726fe69b7069684a973434d1136b03aR70-R72?

File: src/TimeoutMigrationTool.SqlP.FakeData/Program.cs [hunk 16]
```diff
@@ -9,7 +9,7 @@ class Program
     {
         static async Task Main()
         {
-            var endpointConfiguration = new EndpointConfiguration("SqlP.FakeTimeouts");
+            var endpointConfiguration = new EndpointConfiguration("EndpointA");
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool.SqlP.FakeData/Program.cs:12]: Is this change intentional?

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+    using Microsoft.Azure.Cosmos.Table;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using Microsoft.Azure.Cosmos.Table.Queryable;
+
+    public class ASQEndpointMigrator : ITimeoutsTarget.IEndpointTargetBatchMigrator
+    {
+        public ASQEndpointMigrator(CloudTableClient client, string destinationTimeoutTableName, string stagingTableName)
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:13]: Most other sources and targets currently manage the connection internally. Why did you deviate from this here?

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+            this.client = client;
+            this.destinationTimeoutTableName = destinationTimeoutTableName;
+            this.stagingTableName = stagingTableName;
+        }
+
+        public async ValueTask<int> StageBatch(IReadOnlyList<TimeoutData> timeouts, int batchNumber)
+        {
+            await PurgeStagingTable();
+
+            var stagedTimeouts = timeouts.Select(timeout => StagedDelayedMessageEntity.FromTimeoutData(timeout, batchNumber)).GroupBy(x => x.PartitionKey);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:24]: would something like `timeoutsToBeStaged` be a better name?

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+        public ASQEndpointMigrator(CloudTableClient client, string destinationTimeoutTableName, string stagingTableName)
+        {
+            this.client = client;
+            this.destinationTimeoutTableName = destinationTimeoutTableName;
+            this.stagingTableName = stagingTableName;
+        }
+
+        public async ValueTask<int> StageBatch(IReadOnlyList<TimeoutData> timeouts, int batchNumber)
+        {
+            await PurgeStagingTable();
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:22]: As far as I remember purging is not efficient at all when there is large amounts of data in there. In most cases, the recommendation is to delete and recreate the table. There is one caveat do you have to deal with the delete being eventually consistent and when you try to create it right after you need to retry on 409 (conflict)

The doco states though deleting and recreating can take a while https://docs.microsoft.com/en-us/rest/api/storageservices/Delete-Table?redirectedfrom=MSDN#remarks but in practice I've never seen it taking so long. Btw I'm not against sticking to the current way. I think though we should align the approaches in the project because the other branch currently uses delete and recreate

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+            var delayedMessages = await stagingTable.ExecuteQueryAsync(query, int.MaxValue, CancellationToken.None);
+
+            var groupedRecords = delayedMessages.GroupBy(x => x.PartitionKey);
+            await BatchProcessRecords(stagingTable, groupedRecords, (operation, entity) => operation.Delete(entity));
+        }
+
+        async Task PurgeStagingTable()
+        {
+            var table = client.GetTableReference(stagingTableName);
+            var delayedMessages = await table.ExecuteQueryAsync(new TableQuery<DelayedMessageEntity>(), int.MaxValue, CancellationToken.None);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:65]: This has the drawback of materializing ALL the data from the entity to the client while all you really need to delete the entity would be the partition and rowkey

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            var recordsHandled = 0;
+            var currentBatchOperationSize = 0M;
+            var batchOperation = new TableBatchOperation();
+            foreach (var grouping in groupedRecordsByPartition)
+            {
+                var recordsInGroup = grouping.ToList();
+                foreach (var record in recordsInGroup)
+                {
+                    var recordSize = record.CalculateSize();
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:82]: Cool I guess we could introduce something similar on the other PR

File: src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs [hunk 22]
```diff
@@ -0,0 +1,39 @@
+﻿namespace Particular.TimeoutMigrationTool.ASQ
+{
+    using System;
+    using System.Security.Cryptography;
+    using System.Text;
+
+    public class DelayedDeliveryTableNameProvider : IProvideDelayedDeliveryTableName
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs:7]: The interface seems overkill to me. It doesn't look like we really need this extensibility

File: src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+        public DelayedDeliveryTableNameProvider(string overriddenDelayedDeliveryTableName = null)
+        {
+            this.overriddenDelayedDeliveryTableName = overriddenDelayedDeliveryTableName;
+        }
+
+        public string GetDelayedDeliveryTableName(string endpointName)
+        {
+            if (!string.IsNullOrEmpty(overriddenDelayedDeliveryTableName))
+            {
+                return overriddenDelayedDeliveryTableName;
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs:18]: That could create invalid table names, is that intentional?

File: src/TimeoutMigrationTool/ASQ/DelayedMessageEntity.cs [hunk 24]
```diff
@@ -0,0 +1,50 @@
...
+                PartitionKey = GetPartitionKey(timeout.Time),
+                RowKey = $"{GetRawRowKeyPrefix(timeout.Time)}_{timeout.MessageId}",
+                Destination = timeout.Destination,
+                Body = timeout.Body,
+                MessageId = timeout.MessageId,
+                Headers = timeout.Headers
+            };
+        }
+
+        const string PartitionKeyScope = "yyyyMMddHH";
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/DelayedMessageEntity.cs:30]: so that is not configurable like in ASP, yeah 🎆

File: src/TimeoutMigrationTool/Program.cs [hunk 25]
```diff
@@ -652,6 +861,7 @@ static int Main(string[] args)
                     ravenDBCommand.Command("sqlt", ravenDBToSqlTCommand =>
                     {
                         ravenDBToSqlTCommand.Options.Add(targetSqlTConnectionString);
+                        ravenDBToSqlTCommand.Options.Add(targetSqlTSchemaName);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/Program.cs:864]: Nice catch

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 26]
```diff
@@ -0,0 +1,378 @@
...
+    using NUnit.Framework;
+    using Particular.TimeoutMigrationTool;
+    using Particular.TimeoutMigrationTool.ASQ;
+    using System;
+    using System.Collections.Generic;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    [TestFixture]
+    class ASQTargetTests
```
[the remark this answers is not in the document]
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:13]: local editorconfig is the reason. I'll remove configureawaits and add suppress

File: src/TimeoutMigrationTool/ASQ/StagedDelayedMessageEntity.cs [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+    /// <summary>
+    /// Represents a record in the native delays storage table which can be deferred message, saga timeouts, and delayed retries.
+    /// </summary>
+    public class StagedDelayedMessageEntity : TableEntity, ICanCalculateMySize
+    {
+        public string Destination { get; set; }
+        public byte[] Body { get; set; }
+        public string MessageId { get; set; }
+        public string Headers { get; set; }
+        public DateTimeOffset Time { get; set; }
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/StagedDelayedMessageEntity.cs:15]: I seem to remember it is quite dangerous to store DateTimeOffset in table storage due to the automatic conversion that is happening https://stackoverflow.com/questions/28675770/why-doesnt-azure-table-storage-support-datetimeoffset

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs [hunk 28]
```diff
@@ -0,0 +1,378 @@
...
+    {
+        string connectionString = Environment.GetEnvironmentVariable("AzureStorageQueue_ConnectionString") ?? "UseDevelopmentStorage=true;";
+        string endpointName;
+
+        [SetUp]
+        public void Setup()
+        {
+            endpointName = $"TestEndpoint{Guid.NewGuid().ToString().Replace("-", string.Empty)}";
```
[the remark this answers is not in the document]
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:21]: I think we should have a dedicated checkmark in the POA to refactor a lot of this test stuff. Alignment would be one of the things we should do (along with refactor all the hundreds of copy-pasta'd code)

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs (hunk 1 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:65]: Good point.

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs (hunk 2 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:97]: :+1:

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs (hunk 3 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:142]: :+1:

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/DelayedDeliveryTableNameProviderTests.cs (hunk 8 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/DelayedDeliveryTableNameProviderTests.cs:36]: This is a unit test, not touching azure anywhere so I'm not sure it's necessary to be a valid azure table name

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/TimeoutMigrationTool.ASQ.IntegrationTests.csproj (hunk 9 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/TimeoutMigrationTool.ASQ.IntegrationTests.csproj:14]: I'm not sure I understand. Which third party SDK is missing?

File: src/TimeoutMigrationTool.NHibernate.AcceptanceTests/TimeoutMigrationTool.NHibernate.AcceptanceTests.csproj (hunk 10 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.NHibernate.AcceptanceTests/TimeoutMigrationTool.NHibernate.AcceptanceTests.csproj:11]: I'm not sure I understand. Which third party SDK is missing?

File: src/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests.csproj (hunk 11 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests/TimeoutMigrationTool.NHibernate.OracleAcceptanceTests.csproj:20]: I'm not sure I understand. Which third party SDK is missing?

File: src/TimeoutMigrationTool/ASQ/StagedDelayedMessageEntity.cs (hunk 27 above)
Comment (danielmarbach) [reply src/TimeoutMigrationTool/ASQ/StagedDelayedMessageEntity.cs:15]: Ok I guess this is already in use in ASQ so nothing we can do here

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+
+            var timeoutsStaged = await BatchProcessRecords(table, stagedTimeouts, (operation, entity) => operation.InsertOrReplace(entity));
+            return timeoutsStaged;
+        }
+
+        public async ValueTask<int> CompleteBatch(int batchNumber)
+        {
+            var stagingTable = client.GetTableReference(stagingTableName);
+            var query = stagingTable.CreateQuery<StagedDelayedMessageEntity>().Where(x => x.PartitionKey == batchNumber.ToString()).AsTableQuery();
+            var delayedMessages = await stagingTable.ExecuteQueryAsync(query, int.MaxValue, CancellationToken.None);
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:35]: The consequence of using the copied `ExecuteQueryAsync` from ASQ is that we first materialize ALL the data from the server into memory before we execute any other operation. This seems hugely exhaustive and potentially dangerous if many timeouts need to be fetched compared to doing operations per page segment as they stream from the server. That's the reason why we deliberately did not use this method and went for slightly "uglier" code style with the do/while directly in your face where it is needed

**Review by WojcikMike — APPROVED** [review]
I don't have any other comments besides what Daniel already wrote

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs (hunk 7 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:341]: :+1:

File: src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs (hunk 12 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.Raven4.AcceptanceTests/RavenDBToAsqEndToEnd.cs:56]: Yes, I'll add the comment here as well

File: src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPAcceptanceTest.cs (hunk 13 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPAcceptanceTest.cs:33]: Yes, `_` doesn't seem to be liked

File: src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs (hunk 15 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.SqlP.AcceptanceTests/SqlPToASQEndToEnd.cs:109]: Moved

File: src/TimeoutMigrationTool.SqlP.FakeData/Program.cs (hunk 16 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.SqlP.FakeData/Program.cs:12]: Nope

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs (hunk 17 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:13]: The migrator is not the Target, the target already needs a connection. Why should it be created twice?

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs (hunk 19 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:22]: Laila knows everything

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs (hunk 18 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:24]: :+1:

File: src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs (hunk 22 above)
Comment (WilliamBZA) [reply src/TimeoutMigrationTool/ASQ/DelayedDeliveryTableNameProvider.cs:7]: It's only here so that it can be tested :/

File: src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs (hunk 17 above)
Comment (danielmarbach) [reply src/TimeoutMigrationTool/ASQ/ASQEndpointMigrator.cs:13]: Yeah right. Sorry I overlooked that

File: src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs (hunk 28 above)
[the remark this answers is not in the document]
Comment (WilliamBZA) [reply src/TimeoutMigrationTool.ASQ.IntegrationTests/ASQTargetTests.cs:22]: Added to the PoA

**Review by danielmarbach — APPROVED** [review]

**Comment by danielmarbach** [comment]
I'll leave it up to you if you want to squash it or not
